velox/exec/GroupingSet.cpp

```diff
@@ -1081,6 +1095,44 @@ bool GroupingSet::mergeNext(
   VELOX_UNREACHABLE();
 }
 
+bool GroupingSet::mergeNextForDistinct(
+    int32_t maxOutputRows,
+    const RowVectorPtr& result) {
+  VELOX_CHECK(isDistinct());
+
+  bool nextKeyIsEqual{false};
+  bool newDistinct{true};
```

Comment:
Do we need this variable? It is kind of confusing and it doesn't really carry state from one loop iteration to the next.

Reply:
newDistinct is needed. We carry over iterations as we need to determine if a row is distinct or not after reading all the rows with the same key.

Reply:
I might be missing something, but I don't see newDistinct being set to false ever except for stream->id == 0. I think we rely on next.second to be false to detect the last row in a set of duplicate rows and emit it, no?

Reply:
I think I understand now... this is confusing though.

Reply:
I wonder if it makes sense to add a comment explaining this logic, since it is not obvious. My understanding is that we are looping over sorted rows produced by tree-of-loosers. We logically split the stream into runs of duplicate rows. As we process each run we track whether one of the values comes from stream 0, in which case we should not produce a result from that run. Otherwise, we produce a result at the end of the run (when we know for sure whether the run contains a row from stream 0 or not).

Reply:
> I wonder if it makes sense to add a comment explaining this logic, since it is not obvious. My understanding is that we are looping over sorted rows produced by tree-of-loosers. We logically split the stream into runs of duplicate rows. As we process each run we track whether one of the values comes from stream 0, in which case we should not produce a result from that run. Otherwise, we produce a result at the end of the run (when we know for sure whether the run contains a row from stream 0 or not).

@mbasmanova thanks and will update!